src/app/modules/shared_models/member_model.nim

```diff
@@ -278,61 +293,109 @@ QtObject:
       resolvePreferredDisplayName(self.items[ind].localNickname, self.items[ind].ensName, self.items[ind].displayName, self.items[ind].alias) !=
-      resolvePreferredDisplayName(localNickname, ensName, displayName, alias)
-
-    if self.items[ind].displayName != displayName:
-      self.items[ind].displayName = displayName
-      roles.add(ModelRole.DisplayName.int)
+      resolvePreferredDisplayName(localNickname, ensName, displayName, self.items[ind].alias)
 
-    if self.items[ind].ensName != ensName:
-      self.items[ind].ensName = ensName
-      roles.add(ModelRole.EnsName.int)
+    updateRole(displayName, DisplayName)
+    updateRole(ensName, EnsName)
+    updateRole(localNickname, LocalNickname)
+    updateRole(isEnsVerified, IsEnsVerified)
+    updateRole(alias, Alias)
+    updateRole(icon, Icon)
+    updateRole(isContact, IsContact)
+    updateRole(isVerified, IsVerified)
+    updateRole(memberRole, MemberRole)
+    updateRole(joined, Joined)
+    updateRole(isUntrustworthy, IsUntrustworthy)
 
-    if self.items[ind].localNickname != localNickname:
-      self.items[ind].localNickname = localNickname
-      roles.add(ModelRole.LocalNickname.int)
+    if preferredDisplayNameChanged:
+      roles.add(ModelRole.PreferredDisplayName.int)
 
-    if self.items[ind].isEnsVerified != isEnsVerified:
-      self.items[ind].isEnsVerified = isEnsVerified
-      roles.add(ModelRole.IsEnsVerified.int)
+    if roles.len == 0:
+      return
 
-    if self.items[ind].alias != alias:
-      self.items[ind].alias = alias
-      roles.add(ModelRole.Alias.int)
+    if callDataChanged:
+      let index = self.createIndex(ind, 0, nil)
+      defer: index.delete
+      self.dataChanged(index, index, roles)
+
+    return roles
+
+  proc updateItems*(self: Model, items: seq[MemberItem]) =
+    var startIndex = -1
+    var endIndex = -1
+    var allRoles: seq[int] = @[]
+    for item in items:
+      let itemIndex = self.findIndexForMember(item.pubKey)
+      if itemIndex == -1:
+        continue
+      let roles = self.updateItem(
+        item.pubKey,
+        item.displayName,
+        item.ensName,
+        item.isEnsVerified,
+        item.localNickname,
+        item.alias,
+        item.icon,
+        item.isContact,
+        item.isVerified,
+        item.memberRole,
+        item.joined,
+        item.isUntrustworthy,
+        callDataChanged = false,
+      )
+
+      if roles.len > 0:
+        if startIndex == -1:
+          startIndex = itemIndex
+        endIndex = itemIndex
+        allRoles = concat(allRoles, roles)
+
+    if allRoles.len == 0:
+      return
 
-    if self.items[ind].icon != icon:
-      self.items[ind].icon = icon
-      roles.add(ModelRole.Icon.int)
+    let startModelIndex = self.createIndex(startIndex, 0, nil)
+    let endModelIndex = self.createIndex(endIndex, 0, nil)
+    defer: startModelIndex.delete
+    defer: endModelIndex.delete
+    self.dataChanged(startModelIndex, endModelIndex, allRoles)
```

Comment:
The `startModelIndex` and `endModelIndex` should represent the block that has been updated. From what I understand in our case we'll create an update block for items that have not been updated.

Let's consider this case (maybe a good example for the unit test):
```
// current model
[{
   pubKey: "0x1",
   name: "01",
   isContact: false
   },{
   pubKey: "0x2",
   name: "02",
   isContact: false
   },{
   //many more items
   },{
   pubKey: "0x100",
   name: "100",
   isContact: true
   }]
   
   updateItems([{
   pubKey: "0x1",
   name: "01Changed", // Changed
   isContact: true          // Changed
   },{
   pubKey: "0x100",
   name: "0x100",
   isContact: false     // Changed
   }])
```

In this case I think it's optimal to call 2 `dataChanged`. One for the first item with the `name` and `isContact` roleNames and another for the 100th item with `isContact` roleName.

But from what I understand here is that we'll call a single `dataChanged` for all 99 items with `name` and `isContact` roleNames. All these considered IMO we're adding extra complexity here and I'm not sure what will be the benefit.

Reply:
The conclusion here was that's better to emit single `dataChanged` instead of single one per change. We haven't considered anything more complex for now. But hopefully it's good enough for now.

Reply:
Ohh I see now! Didn't expect that 😄 Is it better for this specific model?

In general I'd expect the `dataChanged` to be emitted in a granular manner. Emitting for a block of items and a block of roles would trigger items times roles calls to `data` and qml updates (in the best case scenario). The worst case scenario is when it happens to use proxies with sorters and filters in QML on a role that's being updated.

Reply:
It's not always the best solution. E.g. corner case when every second item is updated. It results with many signal emissions, triggering actions on other model in the chain. When the view receives data change for the whole model it narrows it down to items currently displayed anyways. Even for those items it doesn't mean doing any special operations on ui-side because properties will be detected as not changed if there is no change indeed.

Reply:
Agree on the view. But I'm not concerned about the view. My concern is what happens when the model is used by SFPM.
Let's consider the example in the comment above. We're updating some roles on the first and last item from the model. `name` on the first, `isContact ` on the last. Now if we have a SFPM sorting the contacts by name it will explode 😄. We'll report `name` changed on all model items even though it was updated only on the first item.